Get min hops

rturrado opened the pull request:
The main change of this pull request is in topology.cc, get_min_hops function:
- The `distance + 1` case is now taken into consideration.
- Doc for that function has been improved a lot. Thanks for that, Hans.

The other important change is in mapper.cc:
- I've added a `are_nearest_neighbors` function to encapsulate the `get_min_hops(source, target) == 1` check there. 
- And I've changed the two points in the file doing that check to call that function and look similar in code.
Please have a look at the function's doc and name because none of them may be appropriate.

Apart from this, other minor changes:
- added curly braces for single statement blocks in topology.cc.
- took `source == target` base case out of `connectivity == FULL` at topology.cc, get_distance.
- added a `source == target` base case in topology.cc, get_core_distance.
- added a `source == target` base case in topology.cc, get_min_hops.
- added a GetMinHops test fixture to topology_test.cc.

Commit:
Added num_qubits_per_core as a topology member helper.

Commit:
Added num_qubits_per_core as a topology member helper.

rturrado committed:
Updated nearest neighbors check with an extra check: the two qubits are not communication qubits in two different cores. In that case, they are not mappable.

rturrado committed:
mapper: added are_nearest_neighbors.
topology: updated doc and code for get_min_hops.

rturrado committed:
Updated with Hans' comments.

rturrado committed:
Revert "Added num_qubits_per_core as a topology member helper."

This reverts commit 9e459fa2ed27828a104a2d4a5fb4ba51f3fa1480.

rturrado committed:
Rebased from origin/develop.

rturrado committed:
Added get_min_hops tests.

Commit:
Run tests. Fixed a couple of test cases

rturrado committed:
Merged two if clauses into one.

rturrado committed:
Merge branch 'develop' into get_min_hops

pablolh commented on src/ql/com/topology.cc:
```diff
@@ -685,42 +689,73 @@ utils::UInt Topology::get_distance(Qubit source, Qubit target) const {
 /**
  * Returns the distance between the given two qubits in terms of cores.
  */
 utils::UInt Topology::get_core_distance(Qubit source, Qubit target) const {
```
This should return a bool and be renamed "are_in_same_core" or something like that

rturrado replied:
Yes, good point.

pablolh commented on src/ql/com/topology.cc:
```diff
@@ -685,42 +689,73 @@ utils::UInt Topology::get_distance(Qubit source, Qubit target) const {
+    utils::UInt distance = get_distance(source, target);
+    utils::UInt core_distance = get_core_distance(source, target);
+    QL_ASSERT(core_distance <= distance);
+    if (core_distance == distance) {
```
at this point, distance >= 1 and core_distance is only 0 or 1
so this "if" implies `core_distance == distance == 1` and source and target being in different cores, therefore
`is_comm_qubit(source)` and `is_comm_qubit(target)` are always true, no?

I would rewrite as such, which to me is more readable:

```c++
if (!are_in_same_core(source, target) && is_comm_qubit(source) && is_comm_qubit(target)) {
  QL_ASSERT(distance == 1);
  if (num_comm_qubits == 1) {
    return 3;
  }
  return 2;
}
```
and then `core_distance` variable is useless.

Btw I think we should throw an exception when "edges" key is present or "connectivitiy" is "specified" and num_cores >= 2, in light of what Roberto told me about the assumption with multi-core connectivity.

rturrado replied:
Good. I've updated the function. I've added a comment because, despite the function's documentation, that check can be quite ilegible in some time (so any extra help to the future reader is welcome). Also, I prefer sticking to using `distance + 2` and `distance + 1` instead of `3` and `2` respectively (those `2` and `1` are also magic numbers, but more readable than `3` and `2`, IMO).

rturrado replied:
I have added a `connectivity == SPECIFIED && num_cores >= 2)` check, plus exception, at the `Topology` constructor, right after the `Handle connectivity key` section.

pablolh commented on src/ql/com/topology.cc:
```diff
@@ -685,42 +689,73 @@ utils::UInt Topology::get_distance(Qubit source, Qubit target) const {
- * See the check in initialization of neighbors.
+ * Inside one core, minimum number of hops == distance.
+ *
+ * Between qubits in different cores, the minimum number of hops >= distance + 1.
```
The distinction between "minimum number of hops" and distance is here very confusing.
A distance is a minimum number of hops already.
I think get_min_hops should be renamed with a name that has to do with routing.

rturrado replied:
See my comment below regarding moving `get_min_hops` out of the `Topology` class. I think this change would go together with that other change.

pablolh commented on src/ql/com/topology.cc:
```diff
@@ -685,42 +689,73 @@ utils::UInt Topology::get_distance(Qubit source, Qubit target) const {
  */
 utils::UInt Topology::get_core_distance(Qubit source, Qubit target) const {
-    if (get_core_index(source) == get_core_index(target)) return 0;
+    if (source == target ||
```
If source == target, then logically get_core_index(source) == get_core_index(target), therefore isn't that condition useless? I think previous version was OK no?

rturrado replied:
Well, my point here was that it saves you from calling a couple of functions. But, on the other hand, it is true that it adds an extra check, and it is an extra check that is usually not recommended, like the typical check for self assingment in an assignment operator (we shouldn't check for the unprobable case).

rturrado replied:
I have also removed the assert for `connectivity == FULL` because `get_core_index` already asserts that.

pablolh commented on src/ql/pass/map/qubits/map/detail/mapper.cc:
```diff
@@ -399,6 +399,14 @@ void Mapper::map_routed_gate(const ir::compat::GateRef &gate, Past &past) {
+ * This excludes the case of two communication qubits in different cores and at a distance of exactly one hop.
+ */
+utils::Bool Mapper::are_nearest_neighbors(UInt source, UInt target) const {
+    return platform->topology->get_min_hops(source, target) == 1;
```
Looking again, get_min_hops is only used by the mapper and has to do with routing-specifics.
Maybe it's better to move get_min_hops out of topology and inside mapper. What do you think?

rturrado replied:
I'm with Hans here in that there may be coming changes to Topology in a near future. I would rather keep things as they are at the moment. Even if those changes weren't coming, I would indeed separate this comment's proposal to a different commit. So, let's say, in any case, I would limit the scope of this commit to just fixing the distance cases.

jvansomeren commented:
Yes, given the current all-to-all connectivity of multi-core.
As soon as we deviate from this, get_core_distance comes back.
Having a grid of cores was in my mind when originally writing this.
So I’m ok with rewriting this to a bool but bear in mind that it will come back, probably soon.

Op 3 mei 2023, om 12:19 heeft Pablo Le Hénaff ***@***.******@***.***>> het volgende geschreven:



@pablolh commented on this pull request.

________________________________

In src/ql/com/topology.cc<https://urldefense.com/v3/__https://github.com/QuTech-Delft/OpenQL/pull/478*discussion_r1183465478__;Iw!!PAKc-5URQlI!_9CSc-bfUB6VttmJQnLa-ASjEGdPagFDt9HliDxsHEmlEF8v031McRICXy25XpkOndbx_NVEoJ40DHDcryc8ZrzEyEeyNgz3$>:

> @@ -685,42 +689,73 @@ utils::UInt Topology::get_distance(Qubit source, Qubit target) const {
         }
         return d;
     }
-
     return distance[source][target];
 }

 /**
  * Returns the distance between the given two qubits in terms of cores.
  */
 utils::UInt Topology::get_core_distance(Qubit source, Qubit target) const {


This should return a bool and be renamed "are_in_same_core" or something like that

________________________________

In src/ql/com/topology.cc<https://urldefense.com/v3/__https://github.com/QuTech-Delft/OpenQL/pull/478*discussion_r1183469158__;Iw!!PAKc-5URQlI!_9CSc-bfUB6VttmJQnLa-ASjEGdPagFDt9HliDxsHEmlEF8v031McRICXy25XpkOndbx_NVEoJ40DHDcryc8ZrzEyIWyVHsn$>:

>   */
 utils::UInt Topology::get_min_hops(Qubit source, Qubit target) const {
-    utils::UInt d = get_distance(source, target);
-    utils::UInt cd = get_core_distance(source, target);
-    QL_ASSERT(cd <= d);
-    if (cd == d) {
-        return d+2;
+    if (source == target) {
+        return 0;
+    }
+    utils::UInt distance = get_distance(source, target);
+    utils::UInt core_distance = get_core_distance(source, target);
+    QL_ASSERT(core_distance <= distance);
+    if (core_distance == distance) {


at this point, distance >= 1 and core_distance is only 0 or 1
so this "if" implies core_distance == distance == 1 and source and target being in different cores, therefore
is_comm_qubit(source) and is_comm_qubit(target) are always true, no?

I would rewrite as such, which to me is more readable:

if (!are_in_same_core(source, target) && is_comm_qubit(source) && is_comm_qubit(target)) {
  QL_ASSERT(distance == 1);
  if (num_comm_qubits == 1) {
    return 3;
  } else {
    return 2;
  }
}


and then core_distance variable is useless.

Btw I think we should throw an exception when "edges" key is present or "connectivitiy" is "specified" and num_cores >= 2, in light of what Roberto told me about the assumption with multi-core connectivity.

________________________________

In src/ql/com/topology.cc<https://urldefense.com/v3/__https://github.com/QuTech-Delft/OpenQL/pull/478*discussion_r1183474582__;Iw!!PAKc-5URQlI!_9CSc-bfUB6VttmJQnLa-ASjEGdPagFDt9HliDxsHEmlEF8v031McRICXy25XpkOndbx_NVEoJ40DHDcryc8ZrzEyLK2uAQt$>:

>   *
- * We assume below that a valid path exists with distance+1 hops; this fails
- * when not all qubits in a core support connections to all other cores.
- * See the check in initialization of neighbors.
+ * Inside one core, minimum number of hops == distance.
+ *
+ * Between qubits in different cores, the minimum number of hops >= distance + 1.


The distinction between "minimum number of hops" and distance is here very confusing.
A distance is a minimum number of hops already.
I think get_min_hops should be renamed with a name that has to do with routing.

—
Reply to this email directly, view it on GitHub<https://urldefense.com/v3/__https://github.com/QuTech-Delft/OpenQL/pull/478*pullrequestreview-1410533057__;Iw!!PAKc-5URQlI!_9CSc-bfUB6VttmJQnLa-ASjEGdPagFDt9HliDxsHEmlEF8v031McRICXy25XpkOndbx_NVEoJ40DHDcryc8ZrzEyHV9i4q_$>, or unsubscribe<https://urldefense.com/v3/__https://github.com/notifications/unsubscribe-auth/AEDTBNVF4W6PLYWQ7IDNQF3XEIWLBANCNFSM6AAAAAAXUENLHE__;!!PAKc-5URQlI!_9CSc-bfUB6VttmJQnLa-ASjEGdPagFDt9HliDxsHEmlEF8v031McRICXy25XpkOndbx_NVEoJ40DHDcryc8ZrzEyKM34FCn$>.
You are receiving this because your review was requested.Message ID: ***@***.***>

jvansomeren commented:
The current topology avoids separating intra-core and inter-core connectivities.
Furthermore, I agree that get_min_hops knows about routing in the sense that one of hops must be able to host a 2q gate; that is weird indeed here.
But get_min_hops is also specialized now for multi-core connectivity to be all-to-all.
So this calls for splitting get_min_hops somehow to separate all this.
This splitting can only be sensibly done when we have more topology options.
So I’m fine with anything that works now, but be ready in your mind for new kinds of topologies
and new kinds of routing later.

Op 3 mei 2023, om 12:28 heeft Pablo Le Hénaff ***@***.******@***.***>> het volgende geschreven:



@pablolh commented on this pull request.

________________________________

In src/ql/pass/map/qubits/map/detail/mapper.cc<https://urldefense.com/v3/__https://github.com/QuTech-Delft/OpenQL/pull/478*discussion_r1183509363__;Iw!!PAKc-5URQlI!_629NkbiEDKSm-7Ondxxe7RFy_FPo9zYfIvOiMTwQ8s5IYYXjXqkNR-mQzKNqIVgLzIiGPnLm5zLEZeGwB6Qz9cT125HjMkE$>:

> @@ -399,6 +399,14 @@ void Mapper::map_routed_gate(const ir::compat::GateRef &gate, Past &past) {

 }

+/**
+ * Nearest neighbors are in the same core, and at a distance of exactly one hop.
+ * This excludes the case of two communication qubits in different cores and at a distance of exactly one hop.
+ */
+utils::Bool Mapper::are_nearest_neighbors(UInt source, UInt target) const {
+    return platform->topology->get_min_hops(source, target) == 1;


Looking again, get_min_hops is only used by the mapper and has to do with routing-specifics.
Maybe it's better to move get_min_hops out of topology and inside mapper. What do you think?

—
Reply to this email directly, view it on GitHub<https://urldefense.com/v3/__https://github.com/QuTech-Delft/OpenQL/pull/478*pullrequestreview-1410613613__;Iw!!PAKc-5URQlI!_629NkbiEDKSm-7Ondxxe7RFy_FPo9zYfIvOiMTwQ8s5IYYXjXqkNR-mQzKNqIVgLzIiGPnLm5zLEZeGwB6Qz9cT11dv188y$>, or unsubscribe<https://urldefense.com/v3/__https://github.com/notifications/unsubscribe-auth/AEDTBNXALVPFLIS3GNBRV3DXEIXMJANCNFSM6AAAAAAXUENLHE__;!!PAKc-5URQlI!_629NkbiEDKSm-7Ondxxe7RFy_FPo9zYfIvOiMTwQ8s5IYYXjXqkNR-mQzKNqIVgLzIiGPnLm5zLEZeGwB6Qz9cT1_T3PJWp$>.
You are receiving this because your review was requested.Message ID: ***@***.***>

rturrado committed:
Addressed Pablo's code review's comments.

rturrado committed:
Merge remote-tracking branch 'origin/get_min_hops' into get_min_hops

rturrado committed:
Fixed pytests

rturrado committed:
Removed are_in_same_core, and reused the existing is_inter_core_hop, that is basically the same function, but negated.
Added convenience are_comm_qubits function.
Updated get_min_hops documentation in topology.h.
Renamed local 'distance' variable to 'min_hops', since it was shadowing a Topology member.

rturrado committed:
Merge remote-tracking branch 'origin/get_min_hops' into get_min_hops

pablolh commented on include/ql/com/topology.h:
```diff
@@ -269,6 +269,11 @@ class Topology {
+    /**
+     * Returns whether the two given qubits are communication qubits.
+     */
+    utils::Bool are_comm_qubits(Qubit source, Qubit target) const;
```
Is this function really necessary? I think the && with two is_comm_qubits is fine.

rturrado replied:
Not really. It is just a helper. I would keep it if it makes the code clearer. I've actually just seen another place in the code where it could also be used: `if (is_inter_core_hop(qs, qd) && (!is_comm_qubit(qs) || !is_comm_qubit(qd)))` in `get_neighbors_list` could be rewritten as `if (is_inter_core_hop(qs, qd) && !are_comm_qubit(qs, qd))`. But I leave it up to you to keep it or just use the `is_comm_qubit`.

rturrado committed:
Updated generate_neighbors_list to use are_comm_qubits
